codegen/src/main/java/software/amazon/awssdk/codegen/poet/authscheme/AuthSchemeSpecUtils.java

```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.codegen.poet.authscheme;
+
+import com.squareup.javapoet.ClassName;
+import com.squareup.javapoet.ParameterizedTypeName;
+import com.squareup.javapoet.TypeName;
+import java.util.List;
+import software.amazon.awssdk.codegen.model.intermediate.IntermediateModel;
+import software.amazon.awssdk.codegen.utils.AuthUtils;
+import software.amazon.awssdk.http.auth.spi.HttpAuthOption;
+
+final class AuthSchemeSpecUtils {
+    private final IntermediateModel intermediateModel;
+
+    AuthSchemeSpecUtils(IntermediateModel intermediateModel) {
+        this.intermediateModel = intermediateModel;
+    }
+
+    private String basePackage() {
+        return intermediateModel.getMetadata().getFullAuthSchemePackageName();
+    }
+
+    private String internalPackage() {
+        return intermediateModel.getMetadata().getFullInternalAuthSchemePackageName();
+    }
+
+    public ClassName parametersInterfaceName() {
+        return ClassName.get(basePackage(), intermediateModel.getMetadata().getServiceName() + "AuthSchemeParams");
+    }
+
+    public ClassName parametersInterfaceBuilderInterfaceName() {
+        return parametersInterfaceName().nestedClass("Builder");
+    }
+
+    public ClassName parametersDefaultImplName() {
+        return ClassName.get(internalPackage(), "Default" + parametersInterfaceName().simpleName());
+    }
+
+    public ClassName providerInterfaceName() {
+        return ClassName.get(basePackage(), intermediateModel.getMetadata().getServiceName() + "AuthSchemeProvider");
+    }
+
+    public ClassName providerDefaultImplName() {
+        return ClassName.get(internalPackage(), "Default" + providerInterfaceName().simpleName());
+    }
+
+    public TypeName resolverReturnType() {
+        return ParameterizedTypeName.get(List.class, HttpAuthOption.class);
+    }
+
+    public boolean usesSigV4() {
+        // TODO: reviewer: Does the below method usage change with sigV4a?
```

Comment:
No. The logic of this condition is based on the `signatureVersion` in the service-2.json `metadata` section. When modeling for Sigv4a is introduced, it first of all likely won't affect the `signatureVersion` field, as there are no services that have only Sigv4a today. But in any case, Sigv4a should be added to the internal `AuthType` enum.

Reply:
Yes. My initial thought was the `region` is an auth scheme param for sigv4 and may not be relevant for sigv4a, so something to think about when we support it. Either ways, it is something that will be evident to handle/adjust (if/as needed) when we do sigv4a.